[MRG] ENH: Add option to customize spike histogram colors

ntolley opened the pull request:
As title states.

jasmainak commented on hnn_core/cell_response.py:
```diff
@@ -370,6 +370,13 @@ def plot_spikes_hist(self, trial_idx=None, ax=None, spike_types=None,
+            through colors in the list.
+
+            | Ex: ``'r'``, ``['r', 'g']``
+        If None, default color cycle used.
```
how do we know which color corresponds to which cell type?

ntolley replied:
The legend will indicate the cell type in the plot, but in the code it can be redefined.

jasmainak replied:
I was thinking a dict might be more useful?

```py
dict(evprox='r', evdist='g')
```

rythorpe replied:
Defining a color sequence that gets cycled over is pretty standard for matplotlib and other high-level plotting libraries (e.g., seaborn) is it not? I'm not opposed to using a dictionary, but it's not as much of a lightweight solution as what is implemented now.

Something I just thought of, do we want it to also accept the RGBA color specification as well?

jasmainak replied:
since `net.gid_dict` is a dictionary, it means that the cell types won't always necessarily be in the same order before Python 3.6? Maybe I'm splitting at hairs here, but we should avoid sources of bugs

rythorpe replied:
I don't think that would be a bug. This function currently allows users to specify the cell types, drive types, or grouping of types they wish to plot. If the user makes no underlying assumptions about the order in which the various histograms will be plotted, they can rely on the default behavior which is to plot all cell types based on `net.gid_dict`. If we allow a color sequence to be entered, they could also theoretically use a matplotlib colormap!

ntolley replied:
Would it be a bug that we don't guarantee plots having the same color if python versions change? Also that means that our current code also doesn't have that assurance, so we're not losing anything necessarily

rythorpe commented on hnn_core/cell_response.py:
```diff
@@ -370,6 +370,13 @@ def plot_spikes_hist(self, trial_idx=None, ax=None, spike_types=None,
             | Ex: ``'ev'`` is equivalent to ``['evdist', 'evprox']``
+        color : str | list of str | None
+            String input defining color of histograms plotted. If list of str
+            provided, histograms for each cell type will be plotted by cycling
```
```suggestion
            provided, histograms for each spike type will be plotted by cycling
```
Just to be consistent with our use of the `spike_types` argument.

rythorpe commented on hnn_core/viz.py:
```diff
@@ -353,6 +353,13 @@ def plot_spikes_hist(cell_response, trial_idx=None, ax=None, spike_types=None,
         | Ex: ``'ev'`` is equivalent to ``['evdist', 'evprox']``
+    color : str | list of str | None
+        String input defining color of histograms plotted. If list of str
+        provided, histograms for each cell type will be plotted by cycling
```
```suggestion
        provided, histograms for each spike type will be plotted by cycling
```
Same here.

rythorpe commented:
Aside from my docstring nitpick, LGTM!

jasmainak commented:
okay let's move on if it's good by @chenghuzi !

chenghuzi commented:
> Sounds good! @chenghuzi just to double check is this PR sufficient for your histogram plots in the GUI?

LGTM! I still need to manage the color myself outside of it, but it’s doable.

jasmainak commented:
> I still need to manage the color myself outside of it, but it’s doable.

Can you explain a bit more?

chenghuzi commented:
> > I still need to manage the color myself outside of it, but it’s doable.
> 
> Can you explain a bit more?

So in this case for a single figure, if the user calls the histogram, for example 3 times, the result will be different if we use the `next` iterator to get colors as it's stateful. But we can wrap that operation into the default figure creation function  to make it appear as an all-in-one solution.

jasmainak commented:
is the problem that `color` is used as an iterator to cycle rather than an iterator that has the same length as the number of drives?

ntolley commented:
@rythorpe I implemented that better logic/system to aggregate the spike times in a single list. Also added a bunch of tests to check the edge cases for the dictionary input

jasmainak commented:
@ntolley tests are failing ...

jasmainak commented on hnn_core/cell_response.py:
```diff
@@ -370,6 +370,13 @@ def plot_spikes_hist(self, trial_idx=None, ax=None, spike_types=None,
             Valid strings also include leading characters of spike types
 
             | Ex: ``'ev'`` is equivalent to ``['evdist', 'evprox']``
+        color : str | list of str | None
```
does this documentation need to be updated to match the other file?

jasmainak commented on hnn_core/tests/test_viz.py:
```diff
@@ -161,5 +167,22 @@ def test_dipole_visualization():
+    net.cell_response.plot_spikes_hist(color='r')
+    net.cell_response.plot_spikes_hist(color=['C0', 'C1'])
+    net.cell_response.plot_spikes_hist(color={'beta_prox': 'r',
+                                              'beta_dist': 'g'})
```
Do we have a test that checks that it works fine when there are two drives defined to belong to the same grouping in `spike_types` ?

ntolley replied:
Added!

jasmainak commented:
@chenghuzi does this work for you now with the updates?

ntolley commented on hnn_core/tests/test_viz.py:
```diff
@@ -161,5 +167,22 @@ def test_dipole_visualization():
+    with pytest.raises(TypeError, match="color must be an instance of"):
+        net.cell_response.plot_spikes_hist(color=123)
+    with pytest.raises(ValueError):
+        net.cell_response.plot_spikes_hist(color='z')
```
the warning text changed between versions of matplotlib so there is no match here

ntolley commented on hnn_core/viz.py:
```diff
@@ -353,6 +353,20 @@ def plot_spikes_hist(cell_response, trial_idx=None, ax=None, spike_types=None,
         | Ex: ``'ev'`` is equivalent to ``['evdist', 'evprox']``
+    color : str | dict | None
+        Input defining colors of plotted histograms. If str, all
+        histograms plotted with same color. If list of str provided,
```
fixed docs to match

jasmainak reviewed (approved):
@chenghuzi merge button is yours if you are able to use it in the GUI :)

chenghuzi commented:
> 

LGTM!

chenghuzi commented:
> @chenghuzi merge button is yours if you are able to use it in the GUI :)

Resolved the what's new conflict and now seems everything is good!

jasmainak commented:
@ntolley can you fix the commit history to remove merge commits and resolve the conflicts ?

ntolley commented:
@jasmainak all done! Don't think there are any merge commits? It was just a conflict in whats_new.rst

rythorpe commented:
I'm going to move forward with the rebase and merge. Thanks @ntolley!

jasmainak commented:
amazing thanks @ntolley ! 🥳
